### PR TITLE
[11.0][IMP] web_tree_dynamic_colored_field: more options

### PR DESCRIPTION
**Pull request by jarroyomorales**
* Add attribute ``bg_color_field`` on the tree element's ``colors`` to use as background-color.
* In the tree view declaration, put ``options='{"bg_color_field": "color_field"}`` attribute in the ``field`` tag to have column's background colored in color_field's value.

@etobella @damdam-s @hbrunn @guewen

### CODE REVIEW

**Comment by hbrunn** [comment]
we have https://github.com/OCA/web/pull/1377 already, can you better forward port this?

**Comment by jarroyomorales** [comment]
@hbrunn This module is very different from 10 to 11, and I implemented background-color the same way this module was migrated, so I dont see how forward-porting #1377 would be any different from this. Of course I can change the attribute's name so it is the same in both versions, but do you need something else?

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 1]
```diff
@@ -61,6 +81,8 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
             }
             this.applyColorizeHelper($td, nodeOptions, node, 'fg_color', 'color', ctx);
             this.applyColorizeHelper($td, nodeOptions, node, 'bg_color', 'background-color', ctx);
+            this.applyDirectColorizeHelper($td, nodeOptions, node, 'fg_color_field', 'color', ctx);
```
Comment (hbrunn) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js:108]: this would be `color_field`, right?

**Review by hbrunn — COMMENTED** [review]
yes, just the name please. The code is totally different indeed. But then if I understand the diff correctly, the foreground color never really worked after the migration? It might be better to indeed then make everything uniform as you already did in the code, `bg_color`, `fg_color`, `bg_color_field`, `fg_color_field`, I think this will have least impact. Please update the documentation then accordingly.

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js (hunk 1 above)
Comment (etobella) [reply web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js:108]: No, this is a color_field selected for the column and line (background_color is for all the items of the line).
Sometimes it is hard to create the specifications on a single line and you would like to compute it with python. It is used inside options of the field.

**Comment by jarroyomorales** [comment]
@hbrunn I tested the module and it is all working in 11.0. Do we keep it this way then? I already updated the readme.

**Review by hbrunn — APPROVED** [review]
let's keep it the way it is, yes

**Review by etobella — APPROVED** [review]

**Comment by jarroyomorales** [comment]
This last change allows this module to work with trees that are groupby by default, as it wasnt working properly in previous versions.

**Comment by etobella** [comment]
@hbrunn

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 2]
```diff
@@ -10,21 +10,37 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
          * @override
          */
-        _renderBody: function () {
+        _renderView: function () {
             if (this.arch.attrs.colors) {
-                var colorAttr = this.arch.attrs.colors.split(';')
-                    .filter(color => color.trim().startsWith('color_field'));
+                var colors_attr = this.arch.attrs.colors.split(';')
+                var colorAttr = colors_attr.filter(
+                    attr => attr.trim().startsWith('color_field')
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: Arrow functions are not supported in IE
https://www.odoo.com/documentation/11.0/setup/deploy.html#supported-browsers
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 3]
```diff
@@ -10,21 +10,37 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
+                );
+                var backgroundColorAttr = colors_attr.filter(
+                    attr => attr.trim().startsWith('background_color_field')
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: Arrow functions are not supported in IE
https://www.odoo.com/documentation/11.0/setup/deploy.html#supported-browsers
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

**Review by Tardo — CHANGES_REQUESTED** [review]
Thanks for the efforts!

**Comment by jarroyomorales** [comment]
@Tardo Changes done! Tested on IE11 and worked. Thank you!

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 4]
```diff
@@ -10,21 +10,41 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
          *
          * @override
          */
-        _renderBody: function () {
+        _renderView: function () {
             if (this.arch.attrs.colors) {
-                var colorAttr = this.arch.attrs.colors.split(';')
-                    .filter(color => color.trim().startsWith('color_field'));
+                var colors_attr = this.arch.attrs.colors.split(';')
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: It's not mandatory for the language.... but in OCA we still using semicolon.
```suggestion
                var colors_attr = this.arch.attrs.colors.split(';');
```

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 5]
```diff
@@ -10,21 +10,41 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
+                var colorAttr = colors_attr.filter(
+                    function(attr) {
+                        return attr.trim().lastIndexOf('color_field', 0) === 0
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]:
```suggestion
                        return attr.trim().lastIndexOf('color_field', 0) === 0;
```

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 6]
```diff
@@ -10,21 +10,41 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
+                    }
+                );
+                var backgroundColorAttr = colors_attr.filter(
+                    function(attr) {
+                        return attr.trim().lastIndexOf('background_color_field', 0) === 0
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]:
```suggestion
                        return attr.trim().lastIndexOf('background_color_field', 0) === 0;
```

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 7]
```diff
@@ -33,25 +53,39 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
         _renderBodyCell: function (record, node, colIndex, options) {
             var $td = this._super.apply(this, arguments);
             var ctx = this.getEvalContext(record);
             this.applyColorize($td, record, node, ctx);
             return $td;
         },
 
+        _setDecorationClasses: function (record, $tr) {
+            this._super.apply(this, arguments);
+            this.applyAllColorize($tr, record)
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]:
```suggestion
            this.applyAllColorize($tr, record);
```

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 8]
```diff
@@ -33,25 +53,39 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
             var treeColor = record.data[this.colorField];
             if (treeColor) {
-                $td.css('color', treeColor);
+                $tr.css('color', treeColor);
             }
+            var treeBackgroundColor = record.data[this.backgroundColorField];
+            console.log(this.backgroundColorField)
+            console.log(treeBackgroundColor)
+            if (treeBackgroundColor) {
+                $tr.css('background-color', treeBackgroundColor)
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]:
```suggestion
                $tr.css('background-color', treeBackgroundColor);
```

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 9]
```diff
@@ -33,25 +53,39 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
-        applyColorize: function ($td, record, node, ctx) {
+        applyAllColorize: function ($tr, record) {
             // safely resolve value of `color_field` given in <tree>
             var treeColor = record.data[this.colorField];
             if (treeColor) {
-                $td.css('color', treeColor);
+                $tr.css('color', treeColor);
             }
+            var treeBackgroundColor = record.data[this.backgroundColorField];
+            console.log(this.backgroundColorField)
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: :fire:

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 10]
```diff
@@ -33,25 +53,39 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
+            console.log(treeBackgroundColor)
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: :fire:

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 11]
```diff
@@ -88,6 +124,12 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
             }
         },
 
+        applyDirectColorizeHelper: function ($td, nodeOptions, node, nodeAttribute, cssAttribute, ctx) {
+            var field = nodeOptions[nodeAttribute];
+            if (field && ctx[field])
+                $td.css(cssAttribute, ctx[field]);
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]:
```suggestion
            if (field && ctx[field]) {
                $td.css(cssAttribute, ctx[field]);
            }
```

**Review by Tardo — CHANGES_REQUESTED** [review]
Thanks for the work! only few little comments....

**Comment by jarroyomorales** [comment]
@Tardo Done!

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 12]
```diff
@@ -33,25 +53,37 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
          */
+
         _renderBodyCell: function (record, node, colIndex, options) {
             var $td = this._super.apply(this, arguments);
             var ctx = this.getEvalContext(record);
             this.applyColorize($td, record, node, ctx);
             return $td;
         },
 
+        _setDecorationClasses: function (record, $tr) {
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js:69]: Please, add docstring

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 13]
```diff
@@ -33,25 +53,37 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
+            this._super.apply(this, arguments);
+            this.applyAllColorize($tr, record);
+        },
+
         /**
          * Colorize the current cell depending on expressions provided.
          *
          * @param {Query Node} $td a <td> tag inside a table representing a list view
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: Update this docscript

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 14]
```diff
@@ -33,25 +53,37 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
...
             if (treeColor) {
-                $td.css('color', treeColor);
+                $tr.css('color', treeColor);
             }
+            var treeBackgroundColor = record.data[this.backgroundColorField];
+            if (treeBackgroundColor) {
+                $tr.css('background-color', treeBackgroundColor);
+            }
+        },
+        applyColorize: function ($td, record, node, ctx) {
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js:98]: Add docstring

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 15]
```diff
@@ -88,6 +122,13 @@ odoo.define('web_tree_dynamic_colored_field', function (require) {
             }
         },
 
+        applyDirectColorizeHelper: function ($td, nodeOptions, node, nodeAttribute, cssAttribute, ctx) {
```
Comment (Tardo) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js:144]: Add docstring

**Review by Tardo — CHANGES_REQUESTED** [review]
Looks good! Please, update the docstrings.

**Comment by etobella** [comment]
@Tardo I think all comments have been attended

**Pull request closed without merging**